### PR TITLE
Reformat gardener timestamp annotation

### PR DESCRIPTION
**Pull request by dimityrmirchev**
**How to categorize this PR?**

/area quality
/kind enhancement

**What this PR does / why we need it**:
This PR changes the format of the `gardener.cloud/timestamp` annotation to `time.RFC3339Nano` and unifies the format across different resources. The previously used format is meant for debugging as stated in the [documentation](https://pkg.go.dev/time#Time.String). `Backupbucket` and `Backupentry` were already using `time.RFC3339` format which should be backwards compatible with `time.RFC3339Nano` (see an example [here](https://go.dev/play/p/MiukYosIkkR)).

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
cc @rfranzke 

**Release note**:

```breaking developer
The `gardener.cloud/timestamp` annotation is now formatted as `time.RFC3339Nano`.
```

### CODE REVIEW

File: test/integration/gardenlet/backupbucket/backupbucket_test.go [hunk 1]
```diff
@@ -101,7 +101,7 @@ var _ = Describe("BackupBucket controller tests", func() {
 					ProviderStatus:     providerStatus,
 					LastOperation: &gardencorev1beta1.LastOperation{
 						State:          lastOperationState,
-						LastUpdateTime: metav1.NewTime(fakeClock.Now()),
+						LastUpdateTime: metav1.NewTime(fakeClock.Now().Add(time.Second * 10)),
```
Comment (rfranzke) [inline test/integration/gardenlet/backupbucket/backupbucket_test.go]: Why was this needed?

**Review by rfranzke — COMMENTED** [review]
Thanks @dimityrmirchev for picking this up :)
/cc @timuthy

File: test/integration/gardenlet/backupbucket/backupbucket_test.go (hunk 1 above)
Comment (dimityrmirchev) [reply test/integration/gardenlet/backupbucket/backupbucket_test.go]: Talking about the `Backupbucket` and `Backupentry`, previously the gardener timestamp was rounded to the second (same for the `LastUpdateTime`). Because the gardener timestamp is with greater precision statements involving timestamp comparisons may now return different result. See an example

```golang
if lastUpdateTime.Before(gardenerTimestamp) {
// do something
}
```

This has the potential to return `true` even if `lastUpdateTime` may have happened after `gardenerTimestamp` (because `lastUpdateTime` is rounded and `gardenerTimestamp` is not). In my opinion this is also the case today (correct me if I am wrong). One could argue that the comparison [here](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L191) is not always correct since the reported `lastUpdateTime` may be written by a different node from the one [setting](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L352) the timestamp annotation. Then if one of the nodes is reporting wrong time (imagine a second or two difference between them) we may make false assumptions based on that difference. Perhaps one possible solution may be to replace these comparisons with a comparison to something similar to `metadata.generation`?

**Comment by plkokanov** [comment]
When a `metav1.Time` struct is marshalled, the time is represented as RFC3339 format. Consequently when a string is unmarshalled into a `metav1.Time` string the time is parsed as RCF3339. This means that some information is lost if it was RFC3339Nano previously and you are comparing an annotation field vs a field that is represented as `metav1.Time` (as is the case for the `lastOperation.lastUpdateTime`. I guess this is why you need the changes in the `BackupEntry` and `BackupBucket` reconciler tests, because this check will return true (there is a similar check in both the `BackupBucket` and `BackupEntry` reconcilers): https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L190-L191
I guess this is why lines like this were added: https://github.com/gardener/gardener/blob/d00c246f85eb98edfbfb41f3f79d7662d8af0499/pkg/gardenlet/controller/backupentry/reconciler_test.go#L222

Not sure if we would need the precision of `RFC3339nano` for the checks in the `BackupEntry` and `BackupBucket` reconcilers that I mentioned above

**Comment by dimityrmirchev** [comment]
Thanks @plkokanov for the feedback! Please see my [comment](https://github.com/gardener/gardener/pull/7796#discussion_r1169657143) above. We were anyway losing some precision when formatting the timestamp to `RFC3339`. A simple example:
```golang
func main() {
	one, _ := time.Parse(time.RFC3339, "2023-04-18T08:19:15.1Z")
	two, _ := time.Parse(time.RFC3339, "2023-04-18T08:19:15.2Z")
	if one.Before(two) {
		// we see this printed
		fmt.Println("1")
	}

	one, _ = time.Parse(time.RFC3339, one.Format(time.RFC3339))
	two, _ = time.Parse(time.RFC3339, two.Format(time.RFC3339))

	if one.Before(two) {
		// we want to see this printed but it is not since we lost some precision when formatting
		fmt.Println("2")
	}
}
```

File: pkg/gardenlet/controller/backupbucket/reconciler_test.go [hunk 2]
```diff
@@ -196,6 +197,7 @@ var _ = Describe("Controller", func() {
 	})
 
 	It("should not reconcile the extension BackupBucket if the secret data or extension spec hasn't changed", func() {
+		extensionBackupBucket.Status.LastOperation.LastUpdateTime = metav1.NewTime(time.Now().Add(time.Second * 10).UTC())
```
Comment (timuthy) [inline pkg/gardenlet/controller/backupbucket/reconciler_test.go]:
```suggestion
		extensionBackupBucket.Status.LastOperation.LastUpdateTime = metav1.NewTime(time.Now().Add(time.Second * 1).UTC())
```

Isn't one second enough because that's the maximum rounding error we can have? Also please add a comment because no one will ever remember what happens here 😅

**Review by timuthy — COMMENTED** [review]
I left one question/request in the review, rest looks good to me 🙂

File: test/integration/gardenlet/backupbucket/backupbucket_test.go (hunk 1 above)
Comment (rfranzke) [reply test/integration/gardenlet/backupbucket/backupbucket_test.go]: @shafeeqes Any comment?

Comment (shafeeqes) [reply test/integration/gardenlet/backupbucket/backupbucket_test.go]: Why not use `fakeClock.SetTime(time.Now().Round(time.Second))`, So that the comparison will not be affected?

**Comment by shafeeqes** [comment]
> Not sure if we would need the precision of `RFC3339nano` for the checks in the `BackupEntry` and `BackupBucket` reconcilers that I mentioned above

I agree. 

> One could argue that the comparison [here](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L191) is not always correct since the reported lastUpdateTime may be written by a different node from the one [setting](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L352) the timestamp annotation.

Then, should we round this off before comparing, or is it fine because it's not very probable?

**Comment by plkokanov** [comment]
> > Not sure if we would need the precision of `RFC3339nano` for the checks in the `BackupEntry` and `BackupBucket` reconcilers that I mentioned above
> 
> I agree.
> 
> > One could argue that the comparison [here](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L191) is not always correct since the reported lastUpdateTime may be written by a different node from the one [setting](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L352) the timestamp annotation.
> 
> Then, should we round this off before comparing?


That (rounding off) would be my suggestion as well. However, I wanted to take a deeper look to confirm if there is actually a chance that these timestamps can be less than 1 second a part in practice.

**Comment by plkokanov** [comment]
> > Not sure if we would need the precision of `RFC3339nano` for the checks in the `BackupEntry` and `BackupBucket` reconcilers that I mentioned above
> 
> I agree.
> 
> > One could argue that the comparison [here](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L191) is not always correct since the reported lastUpdateTime may be written by a different node from the one [setting](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L352) the timestamp annotation.
> 
> Then, should we round this off before comparing?


That (rounding off) would be my suggestion as well. However, I wanted to take a deeper look to confirm if there is actually a chance that these timestamps can be less than 1 second a part in practice.

**Comment by dimityrmirchev** [comment]
> > > Not sure if we would need the precision of `RFC3339nano` for the checks in the `BackupEntry` and `BackupBucket` reconcilers that I mentioned above
> > 
> > 
> > I agree.
> > > One could argue that the comparison [here](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L191) is not always correct since the reported lastUpdateTime may be written by a different node from the one [setting](https://github.com/gardener/gardener/blob/c6dab1720910d668f606ddb9362d04c773b7490f/pkg/gardenlet/controller/backupbucket/reconciler.go#L352) the timestamp annotation.
> > 
> > 
> > Then, should we round this off before comparing?
> 
> That (rounding off) would be my suggestion as well. However, I wanted to take a deeper look to confirm if there is actually a chance that these timestamps can be less than 1 second a part in practice.

I think that rounding off before comparing should be fine since it mimics the current behavior. Also there will be no need to modify the tests and add the extra second to the `LastUpdateTime`. PTAL @shafeeqes @plkokanov

File: pkg/gardenlet/controller/backupbucket/reconciler_test.go (hunk 2 above)
Comment (dimityrmirchev) [reply pkg/gardenlet/controller/backupbucket/reconciler_test.go]: One second should be enough, but the rounding off before comparing approach proposed by @shafeeqes and @plkokanov is maybe even better 👀

**Comment by shafeeqes** [comment]
Thanks! Looks good to me.

**Review by timuthy — COMMENTED** [review]
/lgtm

**Comment by dimityrmirchev** [comment]
It seems that the correct function to use should be `Truncate` instead of `Round`. When formatted to `time.RFC3339` the value will not be rounded. See example [here](https://go.dev/play/p/N1FNp6Gpnjc). Also a stress run of the unit test with the change yields:
```
stress -p 16 ./backupentry.test 
5s: 547 runs so far, 0 failures
10s: 1021 runs so far, 0 failures
15s: 1491 runs so far, 0 failures
20s: 1948 runs so far, 0 failures
```
